### PR TITLE
Add support for MELCloud and Heatzy

### PR DESCRIPTION
**Pull request by OlivierZal**

### CODE REVIEW

**Comment by OlivierZal** [comment]
Hi @robertklep, I've made this code change for my own use, I've installed it locally and it seems to work.

If you have some time just to confirm me that I understand well the new logic (but no worries if you don't want).

**Comment by robertklep** [comment]
If I understand it correctly, it creates a single accessory with multiple services (fans, (de)humidifier and window covering). It's nice to see that it works with multiple services that depend on the same required capability (`onoff`), I don't think I actually tested something like that.

From what I can see, it looks okay to me. Passing the accessory to the `onUpdate` is a good idea, it might be useful in other situations too.

**Comment by OlivierZal** [comment]
Thank you very much @robertklep for your feedback.

You understand correctly, MELCloud thermostats allow to support – in addition to traditional heat / cool / auto modes and temperature:
- fan speed
- fan-only and dry modes
- vertical and horizontal vanes direction

IMG_1842

By the way, your new design is much better than the previous one.

The main challenge was to update the display of service statuses "in real time" when switching them on/off directly by pressing the corresponding buttons. Of course the MELCloud app handles the capabilities accordingly in backend, but from what I understand HomeKitty mainly refreshes device capabilities when calling the `getters` and `setters`.
That's why I've chosen to make some temporary settings through the `onUpdate`s.

The only thing weird I see in the log is that the device is regularly deleted and updated:
```
[log] 2022-12-08 15:25:55 [HomeKitty] [EV] device updated — name=Salle de jeux Thermostat id=07656cd0-041c-458b-a54d-7c4ee76e0a49 driver=homey:app:com.mecloud
[log] 2022-12-08 15:25:55 [HomeKitty] - already exposed via HomeKit, will delete to update
[log] 2022-12-08 15:25:55 [HomeKitty] removing device with id 07656cd0-041c-458b-a54d-7c4ee76e0a49 from HomeKit:
[log] 2022-12-08 15:25:55 [HomeKitty] - success 🥳
[log] 2022-12-08 15:25:55 [HomeKitty] [Salle de jeux Thermostat:07656cd0-041c-458b-a54d-7c4ee76e0a49] trying mapper
[log] 2022-12-08 15:25:55 [HomeKitty] [Salle de jeux Thermostat:07656cd0-041c-458b-a54d-7c4ee76e0a49] was able to map 🥳
[log] 2022-12-08 15:25:55 [HomeKitty] [Salle de jeux Thermostat:07656cd0-041c-458b-a54d-7c4ee76e0a49] - device should be exposed
[log] 2022-12-08 15:25:55 [HomeKitty] [Salle de jeux Thermostat - Service.HumidifierDehumidifier, Service.Fan, Service.Fanv2, Service.WindowCovering, Service.Thermostat] creating new service 'HumidifierDehumidifier' for group '' (measure_temperature,onoff,operation_mode,target_temperature,vertical,horizontal,thermostat_mode,fan_power)
[log] 2022-12-08 15:25:55 [HomeKitty] [Salle de jeux Thermostat - Service.HumidifierDehumidifier, Service.Fan, Service.Fanv2, Service.WindowCovering, Service.Thermostat] creating new service 'Fan' for group '' (measure_temperature,onoff,operation_mode,target_temperature,vertical,horizontal,thermostat_mode,fan_power)
[log] 2022-12-08 15:25:55 [HomeKitty] [Salle de jeux Thermostat - Service.HumidifierDehumidifier, Service.Fan, Service.Fanv2, Service.WindowCovering, Service.Thermostat] creating new service 'Fanv2' for group '' (measure_temperature,onoff,operation_mode,target_temperature,vertical,horizontal,thermostat_mode,fan_power)
[log] 2022-12-08 15:25:55 [HomeKitty] [Salle de jeux Thermostat - Service.HumidifierDehumidifier, Service.Fan, Service.Fanv2, Service.WindowCovering, Service.Thermostat] creating new service 'WindowCovering' for group '' (measure_temperature,onoff,operation_mode,target_temperature,vertical,horizontal,thermostat_mode,fan_power)
[log] 2022-12-08 15:25:55 [HomeKitty] [Salle de jeux Thermostat - Service.HumidifierDehumidifier, Service.Fan, Service.Fanv2, Service.WindowCovering, Service.Thermostat] creating new service 'Thermostat' for group '' (measure_temperature,onoff,operation_mode,target_temperature,vertical,horizontal,thermostat_mode,fan_power)
[log] 2022-12-08 15:25:59 [HomeKitty] [EV] device updated — name=Salle de jeux Thermostat id=07656cd0-041c-458b-a54d-7c4ee76e0a49 driver=homey:app:com.mecloud
[log] 2022-12-08 15:25:59 [HomeKitty] - already exposed via HomeKit, will delete to update
[log] 2022-12-08 15:25:59 [HomeKitty] removing device with id 07656cd0-041c-458b-a54d-7c4ee76e0a49 from HomeKit:
[log] 2022-12-08 15:25:59 [HomeKitty] - success 🥳
```

I didn't investigate the reason why yet but will do soon.

**Comment by robertklep** [comment]
@OlivierZal that happens when the device is updated, but it's not clear to me when exactly that event is sent (it's not when a capability it set, but does happen when a device is renamed or capabilities are added/removed, which is why the mapper needs to a delete-then-add so the device runs through the list of maps again). I'm interested to hear if you find out what exactly causes it 😄

And I agree about the design, which is one of the reasons that I chose to create a whole new app. The old "design" was mostly written years ago by others and I never felt like rewriting it.

**Comment by OlivierZal** [comment]
@robertklep, as said today in another thread I found steps to reproduce the debounce case with 100 % success even if the root cause remains obscure. 

Amongst other things I noticed that once we go through https://github.com/robertklep/name.klep.homekitty/blob/bd2ae90f9abce5efaf4ee48df0e5a1d46807d348/app.js#L449 then changes stop being logged in the console (e.g. `this.log` in https://github.com/robertklep/name.klep.homekitty/blob/bd2ae90f9abce5efaf4ee48df0e5a1d46807d348/lib/mapped-device.js#L186) even if capabilities still keep being updated.

**Comment by OlivierZal** [comment]
Hi @robertklep, FYI https://github.com/robertklep/name.klep.homekitty/commit/77cf16cdeeb43d14bcb68d01009c72ef1b50b469 seems to have fixed the listener issue I was talking about in my last comment and also in https://github.com/robertklep/name.klep.homekitty/issues/24#issuecomment-1365887604

File: lib/mapper-accessors.js [hunk 1]
```diff
@@ -111,11 +111,31 @@ module.exports = Mapper => {
...
+            ? value.toUpperCase()
+            : currentTemperature < targetTemperature
+              ? 'HEAT'
+              : currentTemperature > targetTemperature
+                ? 'COOL'
+                : 'OFF';
+          if (currentValue in Characteristic.CurrentHeatingCoolingState) {
+            return Characteristic.CurrentHeatingCoolingState[currentValue];
+          }
+          return Characteristic.CurrentHeatingCoolingState.OFF;
```
Comment (OlivierZal) [inline lib/mapper-accessors.js:127]: @robertklep, FYI I've added some security to HomeKitty's code given that now `thermostat_mode`'s values can be customized with Homey firmware 12.0.1+

File: lib/mapper-accessors.js [hunk 2]
```diff
@@ -111,11 +111,31 @@ module.exports = Mapper => {
...
       },
       Target : {
-        get : value => Characteristic.TargetHeatingCoolingState[ value.toUpperCase() ],
-        set : value => [ 'off', 'heat', 'cool', 'auto' ][value],
+        get : value => {
+          const targetValue = value.toUpperCase();
+          if (targetValue in Characteristic.TargetHeatingCoolingState) {
+            return Characteristic.TargetHeatingCoolingState[targetValue];
+          }
+          return Characteristic.TargetHeatingCoolingState.OFF;
```
Comment (OlivierZal) [inline lib/mapper-accessors.js:136]: @robertklep, same than https://github.com/OlivierZal/name.klep.homekitty/pull/1/files#r1730386545

File: lib/mapper-accessors.js [hunk 3]
```diff
@@ -111,11 +111,31 @@ module.exports = Mapper => {
...
+        },
+        set : value => [ 'off', 'heat', 'cool', 'auto' ][value] ?? 'off',
```
Comment (OlivierZal) [inline lib/mapper-accessors.js:138]: @robertklep, same than https://github.com/OlivierZal/name.klep.homekitty/pull/1/files#r1730386545

**Pull request closed without merging**